core/src/executor/datafusion/datafusion_processor.rs

```diff
@@ -964,4 +1058,262 @@ mod tests {
+            "_data_file_table.sys_hidden_seq_num < _equality_delete_table_0.sys_hidden_seq_num"
+        ));
+
+        // 4. Should have exactly one RIGHT ANTI JOIN
+        assert_eq!(sql.matches("RIGHT ANTI JOIN").count(), 1);
+    }
+
+    /// Test that verifies SQL generation works correctly with only position deletes
+    ///
+    /// This tests the case where we need file path and position columns but not sequence numbers.
+    #[test]
+    fn test_position_deletes_only_hidden_columns() {
+        let project_names = vec!["id".to_owned(), "name".to_owned()];
+        let equality_delete_metadatas = vec![]; // No equality deletes
+
+        // Test scenario: ONLY position deletes (no equality deletes)
+        let builder = SqlBuilder::new(
+            &project_names,
+            Some("_position_delete_table".to_string()),
+            Some("_data_file_table".to_string()),
+            &equality_delete_metadatas,
+            true, // need_file_path_and_pos = true
+        );
+
+        let sql = builder.build_merge_on_read_sql().unwrap();
+
+        // Verify the SQL structure:
+
+        // 1. Should have final SELECT that only returns project columns
+        assert!(sql.starts_with("SELECT id, name FROM ("));
+        assert!(sql.ends_with(") AS final_result"));
+
+        // 2. Inner SELECT should include file path/pos columns but not sequence number
+        assert!(sql.contains(
+            "SELECT id, name, sys_hidden_file_path, sys_hidden_pos FROM _position_delete_table"
+        ));
+        assert!(sql.contains(
+            "SELECT id, name, sys_hidden_file_path, sys_hidden_pos FROM _data_file_table"
+        ));
+        assert!(!sql.contains("sys_hidden_seq_num"));
+
+        // 3. Join conditions should reference file path and position correctly
+        assert!(sql.contains(
+            "_data_file_table.sys_hidden_file_path = _position_delete_table.sys_hidden_file_path"
+        ));
+        assert!(
+            sql.contains("_data_file_table.sys_hidden_pos = _position_delete_table.sys_hidden_pos")
+        );
+
+        // 4. Should have exactly one RIGHT ANTI JOIN
+        assert_eq!(sql.matches("RIGHT ANTI JOIN").count(), 1);
+    }
+
+    /// Test that verifies SQL generation works correctly with no deletes
+    ///
+    /// This is the simplest case - should not add any hidden columns or wrap in final_result.
+    #[test]
+    fn test_no_deletes_no_hidden_columns() {
+        let project_names = vec!["id".to_owned(), "name".to_owned()];
+        let equality_delete_metadatas = vec![]; // No equality deletes
+
+        // Test scenario: NO deletes at all
+        let builder = SqlBuilder::new(
+            &project_names,
+            None, // No position delete table
+            Some("_data_file_table".to_string()),
+            &equality_delete_metadatas,
+            false, // need_file_path_and_pos = false
+        );
+
+        let sql = builder.build_merge_on_read_sql().unwrap();
```

Comment:
Can change the assert to match against the SQL result instead of asserting by `sql.contains`?